spec/20-binary-format.md

```diff
@@ -0,0 +1,106 @@
+out of scope of this specification.
+
+## `Traceparent` binary format
+
+The field `traceparent` encodes the version of the protocol and fields
+`trace-id`, `parent-id` and `trace-flags`. Each field starts with the one byte
+field identifier with the field value following immediately after it. Field
+identifiers are used as markers for additional verification of the value
+consistency and may be used in future for the versioning of the `traceparent`
+field.
+
+``` abnf
+traceparent     = version version_format  
+version         = 1BYTE                   ; version is 0 in the current spec
+version_format  = "{ 0x0 }" trace-id "{ 0x1 }" parent-id "{ 0x2 }" trace-flags
+trace-id        = 16BYTES
+parent-id       = 8BYTES
+trace-flags     = 1BYTE  ; only the least significant bit is used
+```
+
+Unknown field identifier (anything beyond `0`, `1` and `2`) should be treated as
+invalid `traceparent`. All zeroes in `trace-id` and `parent-id` invalidates the
+`traceparent` as well.
+
+## Serialization of `traceparent`
+
+Implementation MUST serialize fields into the field ordering sequence.
+In other words, `trace-id` field should be serialized first, `parent-id`
+second and `trace-flags` - third.
+
+Field identifiers should be treated as unsigned byte numbers and should be
+encoded in big-endian bit order.
+
+Fields `trace-id` and `parent-id` are defined as a byte arrays, NOT a
+long numbers. First element of an array MUST be copied first. When array is
+represented as a memory block of 16 bytes - serialization of `trace-id`
+would be identical to `memcpy` method call on that memory block. This
+may be a concern for implementations casting these fields to integers -
+protocol is NOT defining whether those byte arrays are ordered as big
+endian or little endian and have a sign bit.
+
+If padding of the field is required (`traceparent` needs to be serialized into
+the bigger buffer) - any number of bytes can be appended to the end of the
+serialized value.
+
+## `traceparent` example
+
+``` js
+{0,
+  0, 75, 249, 47, 53, 119, 179, 77, 166, 163, 206, 146, 157, 0, 14, 71, 54,
+  1, 52, 240, 103, 170, 11, 169, 2, 183,
+  2, 1}
+```
+
+This corresponds to:
+
+- `trace-id` is
+  `{75, 249, 47, 53, 119, 179, 77, 166, 163, 206, 146, 157, 0, 14, 71, 54}` or
+  `4bf92f3577b34da6a3ce929d000e4736`.
+- `parent-id` is `{52, 240, 103, 170, 11, 169, 2, 183}` or `34f067aa0ba902b7`.
+- `trace-flags` is `1` with the meaning `recorded` is true.
+
+## `tracestate` binary format
+
+List of up to 32 name-value pairs. Each list member starts with the 1 byte field
+identifier `0`. The format of list member is a single byte key length followed
+by the key value and single byte value length followed by the encoded
+value. Note, single byte length field allows keys and values up to 256
+bytes long. This limit is defined by [trace
+context](https://w3c.github.io/trace-context/#header-value)
+specification. Strings are transmitted in ASCII encoding.
```

Comment:
> Strings are transmitted in ASCII encoding.

Is it a requirement that keys and values must be represented in ASCII? The example below could encode `34f067aa0ba902b7` as 8 bytes instead of 16 ASCII characters, but in this case an instrumentation that transfers the context from text protocol to binary protocol would need to know somehow which vendor values are actually hexadecimal representations, which doesn't sound like something we want.

If we say that keys/values must indeed be ASCII strings, then it makes me question why we need binary encoding at all, since it does not provide any compression. It might be slightly faster to parse since the sizes are explicit.

Reply:
I was thinking about it. See https://github.com/w3c/trace-context-protocols-registry/issues/4#issuecomment-461916686 

> ## Charset of `tracestate` value

> HTTP-imposed limitation on character set of the opaque value of `tracestate` value doesn't may not be an issue for other protocols, which generally support `UTF-8` charset. This potentially may lead to various issues like one may assume UTF-8 charset for messaging protocol and will assume something like `urlencoding` will be applied to non-ASCII characters. However specification operates with opaque values and does not assume any encoding. So value may be simply dropped by implementation.

> One of the way to solve it is in fact suggest some encoding mechanism specifically for HTTP protocol. This will, however, complicate implementations and is hard to test in existing test harness (it only supports binary now).

We can compress by utilizing space of non-printeable characters in ASCII table. It doesn't feel however we it's a lot of characters. And there is no commonly used encoding I know that will help do this. So we will need to invent a new one.

As you pointed we can simply do `tracestate` as a string as ASCII then. Since we do almost no compression. The only problem with this from my standpoint is the future proofing for versioning of `tracestate` fields which is provided by field ids. We can introduce compression later using by versioning field ID.

Also we most probably will use a similar format for `correlation-context`. Similar format already used in GRPC by Open Census. So it will not be completely new code to parse this field.

Do you have a strong preference one way or another? Should we look closer into compression now from your perspective?

Reply:
Created an issue for this: https://github.com/w3c/trace-context-binary/issues/4